File: internal/control/validate_test.go
package control

import "testing"

func TestValidNodeName(t *testing.T) {
	for _, name := range []string{"home-nas", "家里 NAS", "n1", "studio.local"} {
		if err := validNodeName(name); err != nil {
			t.Errorf("%q: %v", name, err)
		}
	}
	for _, name := range []string{"", "   ", "!!! bad/name?", "---", "a/b", "no?"} {
		if err := validNodeName(name); err == nil {
			t.Errorf("%q: accepted", name)
		}
	}
}

func TestValidLocalHost(t *testing.T) {
	for _, host := range []string{"127.0.0.1", "localhost", "nas.local", "::1", "[::1]", "my_nas"} {
		if !validLocalHost(host) {
			t.Errorf("%q: rejected", host)
		}
	}
	for _, host := range []string{"", "not_a_host!!!", "host name", "a/b"} {
		if validLocalHost(host) {
			t.Errorf("%q: accepted", host)
		}
	}
}

func TestValidCidrsExamples(t *testing.T) {
	if err := validateCidrs(""); err != nil {
		t.Fatal(err)
	}
	if err := validateCidrs("10.0.0.0/8, 192.168.1.1"); err != nil {
		t.Fatal(err)
	}
	if err := validateCidrs("not-a-cidr"); err == nil {
		t.Fatal("not-a-cidr accepted")
	}
}

File: src/lib/umbra/validate.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { validCidrs, validHost, validNodeName } from "./validate.ts";

test("node names reject punctuation and path characters", () => {
  assert.equal(validNodeName("home-nas"), true);
  assert.equal(validNodeName("家里 NAS"), true);
  assert.equal(validNodeName("n1"), true);
  assert.equal(validNodeName("studio.local"), true);
  assert.equal(validNodeName("!!! bad/name?"), false);
  assert.equal(validNodeName(""), false);
  assert.equal(validNodeName("   "), false);
  assert.equal(validNodeName("---"), false);
});

test("target hosts accept IP and hostname, reject garbage", () => {
  assert.equal(validHost("127.0.0.1"), true);
  assert.equal(validHost("localhost"), true);
  assert.equal(validHost("nas.local"), true);
  assert.equal(validHost("::1"), true);
  assert.equal(validHost("[::1]"), true);
  assert.equal(validHost("my_nas"), true);
  assert.equal(validHost("not_a_host!!!"), false);
  assert.equal(validHost("host name"), false);
  assert.equal(validHost(""), false);
});

test("CIDR allowlists accept empty, IP, and prefixes", () => {
  assert.equal(validCidrs(""), true);
  assert.equal(validCidrs("10.0.0.0/8"), true);
  assert.equal(validCidrs("192.168.1.1"), true);
  assert.equal(validCidrs("10.0.0.0/8, 192.168.0.0/16"), true);
  assert.equal(validCidrs("2001:db8::/32"), true);
  assert.equal(validCidrs("not-a-cidr"), false);
  assert.equal(validCidrs("10.0.0.0/33"), false);
});
